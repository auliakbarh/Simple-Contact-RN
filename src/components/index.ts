// atom
export * from '@/components/atom/Typography'
export * from '@/components/atom/ShimmerPlaceholder'

// molecule
export * from '@/components/molecule/Header'
export * from '@/components/molecule/form/InputFieldText'

// wrapper
export * from '@/components/atom/wrapper/SafeArea'
export * from '@/components/atom/wrapper/AppWrapper'

