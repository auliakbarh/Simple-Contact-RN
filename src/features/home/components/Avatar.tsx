import React, {useMemo, useState} from 'react';
import {View,Image} from 'react-native';

import {isBlank} from "@/utils";
import {tw} from "@/config";
import {Typography} from "@/components/atom/Typography";

export interface IAvatarProps {
  firstName: string
  lastName: string
  photoUrl?: string | null
}

function getInitials (first: string, last: string): string {
  const words = [first, last]
  const initials = (words.slice(0, 2)).map(word => !isBlank(word) ? word.charAt(0) : '')

  return initials.join('')
}

export const Avatar: React.FC<IAvatarProps> = ({firstName, lastName, photoUrl}) => {
  const initials = useMemo(() => {
    return getInitials(firstName ?? '#', lastName ?? '#')
  }, [firstName, lastName])

  const styles = useMemo(() => [tw`w-[200px] bg-[#bdbdbd] rounded-full center`, { aspectRatio: 1 }], [])

  const [uri, setUri] = useState<string | null | undefined>(photoUrl)

  const src = useMemo(() => {
    if (uri) {
      return uri
    }

    return undefined
  }, [uri])

  return (
    src !== undefined
      ? <Image src={src} style={styles} onError={() => { setUri(undefined) }}/>
      : <View style={styles}>
        <Typography style={tw`text-[72px] font-bold`}>{initials}</Typography>
      </View>
  )
}
