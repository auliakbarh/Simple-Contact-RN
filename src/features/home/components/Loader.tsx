import React from 'react';
import {View} from "react-native";

import {ShimmerPlaceholder} from "@/components/atom/ShimmerPlaceholder";
import {tw} from "@/config";

interface ILoaderProps extends React.PropsWithChildren {
  loading: boolean;
}

export const Loader: React.FC<ILoaderProps> = ({children, loading}) => {
  if (!loading) {
    return children
  }

  return (
    <View style={tw`p-4`}>
      <ShimmerPlaceholder visible={false} style={tw`h-50 w-full rounded-4`} />
      <View style={tw`h-40 pt-4`}>
        <View style={tw`h-[1px] bg-black dark:bg-neutral w-10 self-center`}/>
      </View>
    </View>
  )
}
