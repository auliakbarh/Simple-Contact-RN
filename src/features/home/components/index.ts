export * from '@/features/home/components/BottomSheetDetailContact'
