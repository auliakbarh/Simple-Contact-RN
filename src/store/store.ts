import {TypedUseSelectorHook, useDispatch, useSelector} from "react-redux";
import {configureStore} from '@reduxjs/toolkit'
import {settingSlice} from '@/store/slice/setting'
import {contactService} from "@/store/service/contact";

export const store = configureStore({
  reducer: {
    [settingSlice.reducerPath]: settingSlice.reducer,
    [contactService.reducerPath]: contactService.reducer,
  },
  middleware: getDefaultMiddleware => getDefaultMiddleware().concat(contactService.middleware)
})

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch

export const useAppDispatch: () => AppDispatch = useDispatch;
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
