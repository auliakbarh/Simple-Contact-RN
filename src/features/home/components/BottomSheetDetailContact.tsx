import React, {useImperativeHandle, useMemo, useRef, useState} from 'react'
import {ScrollView, TouchableOpacity, View} from "react-native";
import {BottomSheetModalMethods} from "@gorhom/bottom-sheet/lib/typescript/types";
import {BottomSheetBackdrop, BottomSheetModal, BottomSheetView} from "@gorhom/bottom-sheet";
import {tw} from "@/config";
import {Typography} from "@/components/atom/Typography";
import {useTranslation} from "react-i18next";
import {Ionicons} from "@expo/vector-icons";
import {useColor} from "@/hooks";
import {Avatar} from "@/features/home/components/Avatar";
import {isBlank} from "@/utils";

export interface IBottomSheetDetailContactProps {
  data?: IContact
  id?: string
}

export interface IBottomSheetDetailContactMethod extends Partial<BottomSheetModalMethods> {}

export const BottomSheetDetailContact = React.forwardRef<IBottomSheetDetailContactMethod, IBottomSheetDetailContactProps>(({ data, id }, ref) => {
  const bottomSheetModalRef = useRef<BottomSheetModal>(null)
  const snapPoints = useMemo(() => ['66%', '90%'], [])
  const {t} = useTranslation('messages');
  const {t: tFields} = useTranslation('add-new-contact');

  const [state, setState] = useState<IContact | undefined>()

  const detail = useMemo(() => {
    if (state) {
      return state
    }

    if (data) {
      return data
    }
    return null
  }, [data, state])

  const colorFill = useColor('fill')

  const fields = useRef({ firstName: tFields('first-name'), lastName: tFields('last-name'), age: tFields('age') }).current

  useImperativeHandle(ref, () => ({
    ...bottomSheetModalRef.current,
    present: (val: IContact) => {
      bottomSheetModalRef.current?.present()
      setState(val)
    },
    close: (animationConfigs) => {
      bottomSheetModalRef.current?.close(animationConfigs)
      setState(undefined)
    },
    forceClose: (animationConfigs) => {
      bottomSheetModalRef.current?.forceClose(animationConfigs)
      setState(undefined)
    },
    dismiss: (animationConfigs) => {
      bottomSheetModalRef.current?.dismiss(animationConfigs)
      setState(undefined)
    },
    collapse: (animationConfigs) => {
      bottomSheetModalRef.current?.collapse(animationConfigs)
      setState(undefined)
    },
  }))

  return (
    <BottomSheetModal
      ref={bottomSheetModalRef}
      name="bottom-sheet-detail-contact"
      index={1}
      snapPoints={snapPoints}
      handleStyle={tw`bg-neutral dark:bg-black py-5 rounded-t-4`}
      handleIndicatorStyle={tw`bg-black dark:bg-neutral`}
      backdropComponent={(backdropProps) => (
        <BottomSheetBackdrop {...backdropProps} enableTouchThrough={true}/>
      )}>

      {detail != null
        ? (
          <BottomSheetView style={tw`fill items-center bg-neutral dark:bg-night`}>
            <View style={tw`fill w-full`}>
              <View style={tw`absolute left-4 right-4 top-4 flex-1 row justify-between items-center z-10`}>
                <TouchableOpacity
                  onPress={() => bottomSheetModalRef.current?.dismiss()}
                  testID="close-modal-button">
                  <Ionicons name="chevron-back" size={48} color={colorFill}/>
                </TouchableOpacity>
              </View>

              <ScrollView style={tw`w-full`}>
                <View style={tw`mt-6 items-center`}>
                  <Avatar firstName={detail.firstName} lastName={detail.lastName} photoUrl={detail.photo} />
                  <Typography style={tw`mt-6 text-h1`}>{detail.firstName + (!isBlank(detail.lastName ?? '') ? ` ${detail.lastName}` : '')}</Typography>
                </View>
                <View style={tw`mt-8 px-6`}>
                  {
                    Object.keys(fields).map((key, index) => {
                      return (
                        <React.Fragment key={key}>
                          <View style={tw`row items-center`}>
                            <Typography style={tw`text-h-3 font-bold flex-1 mr-6`}>{fields[key as keyof typeof fields]}:</Typography>
                            <Typography style={tw`text-h-3 flex-3`}>{detail[key as keyof IContact]}</Typography>
                          </View>
                          {index < Object.keys(fields).length - 1 && (<View style={tw`h-5`} />)}
                        </React.Fragment>
                      )
                    })
                  }
                </View>
              </ScrollView>
            </View>
          </BottomSheetView>
        )
        : (
          <View style={tw`fill center bg-neutral dark:bg-night`}>
            <Typography>{t('error-invalid-contact')}</Typography>
          </View>
        )
      }

    </BottomSheetModal>
  )
})
