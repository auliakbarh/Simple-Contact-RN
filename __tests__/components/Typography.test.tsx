import React from 'react'
import {Text} from 'react-native'
import {Provider} from 'react-redux';
import Renderer from 'react-test-renderer'
import {Typography} from "@/components"
import {store} from "@/store/store";

describe('<Typography />', () => {
  it('renders correctly', () => {
    const tree = Renderer.create(
      <Provider store={store}>
        <Typography>Aulia Akbar Harahap</Typography>
      </Provider>
    )

    const root = tree.root
    const typography = root.findByType(Text)

    expect(typography.props.children).toBeDefined()
    expect(typography.props.children).toBe('Aulia Akbar Harahap')
    expect(tree).toMatchSnapshot()
  })
})
