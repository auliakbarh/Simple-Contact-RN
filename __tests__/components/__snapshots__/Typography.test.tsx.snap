// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Typography /> renders correctly 1`] = `
<Text
  style={
    [
      {
        "color": "#111111",
        "fontSize": 16,
      },
      undefined,
    ]
  }
>
  Aulia Akbar Harahap
</Text>
`;
